File: infra/lambda/page-renderer/notion_renderer.py
"""Notion 공개 페이지의 본문 준비 상태와 접이식 블록을 처리한다."""

from urllib.parse import urlparse

from playwright.sync_api import Error as PlaywrightError
from playwright.sync_api import TimeoutError as PlaywrightTimeoutError

NOTION_DOMAINS = ("notion.com", "notion.site")
CONTENT_WAIT_MS = 10_000
PAGE_SETTLE_WAIT_MS = 2_000
TOGGLE_CLICK_WAIT_MS = 500
MAX_TOGGLE_CLICKS = 100
CLOSED_TOGGLE_SELECTOR = (
    '.notion-page-content .notion-selectable.notion-toggle-block '
    '[role="button"][aria-expanded="false"], '
    '.notion-page-content .notion-selectable.notion-header-block '
    '[role="button"][aria-expanded="false"], '
    '.notion-page-content .notion-selectable.notion-sub_header-block '
    '[role="button"][aria-expanded="false"], '
    '.notion-page-content .notion-selectable.notion-sub_sub_header-block '
    '[role="button"][aria-expanded="false"]'
)


def is_notion_url(url):
    try:
        host = (urlparse(url).hostname or "").lower()
    except ValueError:
        return False

    return any(
        host == domain or host.endswith(f".{domain}")
        for domain in NOTION_DOMAINS
    )


def wait_for_content(page):
    """Notion 앱 셸이 아니라 실제 문서 블록이 렌더링될 때까지 기다린다."""
    try:
        page.wait_for_function(
            """
            () => {
                const content = document.querySelector('.notion-page-content');
                if (!content || !content.innerText.trim()) {
                    return false;
                }

                return content.querySelector('.notion-selectable') !== null;
            }
            """,
            timeout=CONTENT_WAIT_MS,
        )
        return True
    except PlaywrightTimeoutError:
        print(f"[render] notion content wait timeout url={page.url}")
        return False


def settle(page):
    page.wait_for_timeout(PAGE_SETTLE_WAIT_MS)


def expand_toggles(page):
    """Notion 본문의 닫힌 toggle과 접이식 heading을 모두 연다."""
    expanded_count = 0
    for _ in range(MAX_TOGGLE_CLICKS):
        toggles = page.locator(CLOSED_TOGGLE_SELECTOR)
        before_count = toggles.count()
        if before_count == 0:
            break

        toggle = toggles.nth(0)
        if not toggle.is_visible():
            print("[render] notion toggle skipped reason=not-visible")
            break

        try:
            # Locator는 DOM 변경 후 재매칭되므로 클릭 대상의 handle을 보존한다.
            toggle_handle = toggle.element_handle()
            toggle.scroll_into_view_if_needed()
            toggle.click(timeout=2_000, force=True)
            page.wait_for_timeout(TOGGLE_CLICK_WAIT_MS)
        except PlaywrightError as exception:
            print(f"[render] notion toggle click failed reason={exception}")
            break

        handle_expanded = (
            toggle_handle is not None
            and toggle_handle.get_attribute("aria-expanded") == "true"
        )
        after_count = page.locator(CLOSED_TOGGLE_SELECTOR).count()
        if not handle_expanded and after_count >= before_count:
            print(
                "[render] notion toggle click had no effect "
                f"collapsed={before_count}->{after_count}"
            )
            break

        expanded_count += 1

    remaining = page.locator(CLOSED_TOGGLE_SELECTOR).count()
    print(
        "[render] notion toggle expansion finished "
        f"expanded={expanded_count} remaining={remaining}"
    )

File: infra/lambda/page-renderer-node/notion_renderer.mjs
const CONTENT_WAIT_MS = 10_000;
const PAGE_SETTLE_WAIT_MS = 2_000;
const TOGGLE_CLICK_WAIT_MS = 500;
const MAX_TOGGLE_CLICKS = 100;

export function isNotionUrl(url) {
  try {
    const host = new URL(url).hostname.toLowerCase();
    return host === 'notion.com' || host.endsWith('.notion.com')
      || host === 'notion.site' || host.endsWith('.notion.site');
  } catch {
    return false;
  }
}

export async function prepare(page) {
  try {
    await page.waitForFunction(() => {
      const content = document.querySelector('.notion-page-content');
      return Boolean(content?.innerText.trim() && content.querySelector('.notion-selectable'));
    }, { timeout: CONTENT_WAIT_MS });
  } catch (error) {
    console.log(JSON.stringify({ stage: 'notion-content-unavailable', message: error.message }));
    return false;
  }

  await page.waitForTimeout(PAGE_SETTLE_WAIT_MS);
  await expandToggles(page);
  return true;
}

async function expandToggles(page) {
  const selector = [
    '.notion-page-content .notion-selectable.notion-toggle-block [role="button"][aria-expanded="false"]',
    '.notion-page-content .notion-header-block [role="button"][aria-expanded="false"]',
    '.notion-page-content .notion-sub_header-block [role="button"][aria-expanded="false"]',
    '.notion-page-content .notion-sub_sub_header-block [role="button"][aria-expanded="false"]',
  ].join(', ');

  let expanded = 0;
  for (let index = 0; index < MAX_TOGGLE_CLICKS; index += 1) {
    const toggles = page.locator(selector);
    const before = await toggles.count();
    if (before === 0) break;

    const toggle = toggles.first();
    if (!await toggle.isVisible()) break;
    try {
      await toggle.scrollIntoViewIfNeeded();
      await toggle.click({ timeout: 2_000, force: true });
      await page.waitForTimeout(TOGGLE_CLICK_WAIT_MS);
      expanded += 1;
    } catch (error) {
      console.log(JSON.stringify({ stage: 'notion-toggle-click-failed', message: error.message }));
      break;
    }
  }

  console.log(JSON.stringify({
    stage: 'notion-toggle-expansion-finished',
    expanded,
    remaining: await page.locator(selector).count(),
  }));
}
